### PR TITLE
openjdk 22, openjdk@21 21.0.2 (new formula)

### PR DESCRIPTION
**Pull request by WangEdward**
Prepare for moving to OpenJDK 22 #166788

### CODE REVIEW

**Comment by ZhongRuoyu** [comment]
Thanks @WangEdward, can you do this (and all the migrations to `openjdk@21`) in #166788 instead? Otherwise we'd have two identical `openjdk` formulae before #166788 gets merged.

**Comment by WangEdward** [comment]
@chenrui333 @ZhongRuoyu 
Commits in #166788 are too messy, so I moved everything here. Could you please add long build and CI-long-timeout for this. Thanks!

**Comment by ZhongRuoyu** [comment]
Squashed b749377d02f0dc11a5f839c22aab41b05789d884 and 9f5cb0a831287c58a6e8e94f055b2d155da87f98.

**Review by ZhongRuoyu — COMMENTED** [review]
These changes should help to fix the build failures. But please avoid pushing until CI finishes because it takes quite a long time to do one run. Thanks!

File: Formula/s/scm-manager.rb
Comment (ZhongRuoyu) [inline Formula/s/scm-manager.rb:17]: Don't need to mention this; CI will explain.

```suggestion
```

**Review by ZhongRuoyu — COMMENTED** [review]
@WangEdward, we seek your cooperation by avoid pushing to this branch before CI is completed. Due to https://github.com/Homebrew/homebrew-core/labels/CI-build-dependents-from-source, the tests take a very long time and we don't want to start over after hours of run like this: https://github.com/Homebrew/homebrew-core/actions/runs/8602702349/job/23584693893?pr=168294.

**Comment by WangEdward** [comment]
> @WangEdward, we seek your cooperation by avoid pushing to this branch before CI is completed. Due to [ CI-build-dependents-from-source](https://github.com/Homebrew/homebrew-core/labels/CI-build-dependents-from-source) Pass --build-dependents-from-source to brew test-bot. , the tests take a very long time and we don't want to start over after hours of run like this: https://github.com/Homebrew/homebrew-core/actions/runs/8602702349/job/23584693893?pr=168294.

Sorry about that. I will make sure to push after all jobs are finished next time. 

I thought that the test results were likely to be the same for the arm and intel machine, and if there are any extra failing items for intel, it's not likely to be an OpenJDK-related problem. The build time on intel machines is extensively long (>10 hrs), and if we are eventually going to run the whole test for all the machines, why bother wasting time on testing it at the first build?

**Comment by ZhongRuoyu** [comment]
> I thought that the test results were likely to be the same for arm and intel machines, and if there are any extra failing items for intel, it's not likely to be an OpenJDK-related problem.

Although that's likely the case in general, it also comes with the assumption that the OpenJDK build itself performs as expected. We are usually more cautious with compilers, considering how they might affect users and also the dependent formulae. (And OpenJDK is a bit special in that formulae built with it also depend on it at runtime.)

> The build time on intel machines is extensively long (>10 hrs), and if we are eventually going to run the whole test for all the machines, why bother wasting time on testing it at the first build?

Because subsequent tests will not be run with https://github.com/Homebrew/homebrew-core/labels/CI-build-dependents-from-source, once we have a complete list of what's broken.

**Comment by landsman** [comment]
Guys, can you please create an issue to describe what has to be done to introduce openjdk@22? I'm a bit lost in this communication.

**Comment by ZhongRuoyu** [comment]
At this moment all we need is patience. Many PRs are waiting for a long-timeout CI slot and we'll rerun CI here when the backlog is lighter.

**Comment by calvinit** [comment]
openjdk@21 [21.0.3](https://www.oracle.com/java/technologies/javase/21-0-3-relnotes.html) has been released. Can we update this formula first? 22 non-LTS, I think can be dealt with later.

PS: [JDK Releases](https://www.java.com/releases/).

**Comment by WangEdward** [comment]
> openjdk@21 [21.0.3](https://www.oracle.com/java/technologies/javase/21-0-3-relnotes.html) has been released. Can we update this formula first? 22 non-LTS, I think can be dealt with later.
> 
> PS: [JDK Releases](https://www.java.com/releases/).

You can open a pull request to bump openjdk to 21.0.3 if you want to. I will rebase this commit afterward.

**Comment by jfallows** [comment]
Guys, is this ready to make further progress now? 🙏

**Comment by SMillerDev** [comment]
Build error for jsvc on macOS 14 (ARM). Probably due to an XCode version. Let's run a rebottle to check.

```console
  jsvc-unix.c:1311:20: error: incompatible function pointer types assigning to 'void (*)(int)' from 'void (int, siginfo_t *, void *)' (aka 'void (int, struct __siginfo *, void *)') [-Wincompatible-function-pointer-types]
      act.sa_handler = controller;
                     ^ ~~~~~~~~~~
  1 error generated.
  make[1]: *** [jsvc-unix.o] Error 1
  make[1]: *** Waiting for unfinished jobs....
```

**Comment by WangEdward** [comment]
> Build error for jsvc on macOS 14 (ARM). Probably due to an XCode version. Let's run a rebottle to check.

It's possible that `jsvc` does not support macOS 14 as the [latest version](https://github.com/Homebrew/homebrew-core/pull/130856/files) was released prior to the Sonoma release.

**Comment by SMillerDev** [comment]
Yeah, that rebottle confirms it. Could you try and make an issue upstream about it?

**Comment by SMillerDev** [comment]
If the x86 runners pass (except jsvc) we can merge this one IYAM

**Comment by SMillerDev** [comment]
Created https://issues.apache.org/jira/browse/DAEMON-463

**Comment by michael-o** [comment]
> Build error for jsvc on macOS 14 (ARM). Probably due to an XCode version. Let's run a rebottle to check.
> 
> ```
>   jsvc-unix.c:1311:20: error: incompatible function pointer types assigning to 'void (*)(int)' from 'void (int, siginfo_t *, void *)' (aka 'void (int, struct __siginfo *, void *)') [-Wincompatible-function-pointer-types]
>       act.sa_handler = controller;
>                      ^ ~~~~~~~~~~
>   1 error generated.
>   make[1]: *** [jsvc-unix.o] Error 1
>   make[1]: *** Waiting for unfinished jobs....
> ```

I recommend to apply temporary fix, the same which FreeBSD has applied until resolved upstream: https://github.com/freebsd/freebsd-ports/blob/e8c3e827f61838b8bd8509e3d4a6d39dfbadfdd1/devel/apache-commons-daemon/Makefile#L32-L34

**Comment by WangEdward** [comment]
Full CI run: https://github.com/Homebrew/homebrew-core/actions/runs/8949785914

I think we no longer need to build dependents, as there were no dependent failures related to OpenJDK in the last CI check.

**Comment by SMillerDev** [comment]
@WangEdward can you make that JSVC fix a separate PR so we can rebase on top of that here after?

**Review by SMillerDev — APPROVED** [review]

**Comment by calvinit** [comment]
Hi, guys. Is there anything we can do next? @everyone

**Comment by iMichka** [comment]
I have a fix for `cbmc`: #172959

**Comment by ZhongRuoyu** [comment]
`carrot2` build seems stuck on {13,14}-x86_64.

**Comment by iMichka** [comment]
Requed to see if the `carrot2` issue still happens.

**Review by ZhongRuoyu — APPROVED** [review]
Thanks @WangEdward @SMillerDev @iMichka and everyone!

**Comment by calvinit** [comment]
> > openjdk@21 [21.0.3](https://www.oracle.com/java/technologies/javase/21-0-3-relnotes.html) has been released. Can we update this formula first? 22 non-LTS, I think can be dealt with later.
> > PS: [JDK Releases](https://www.java.com/releases/).
> 
> #170043

@WangEdward Hi guy, now this PR has been merged. Would you mind adding a new PR to update the latest `openjdk` **22.0.1**? Thanks.

**Comment by ZhongRuoyu** [comment]
> Would you mind adding a new PR to update the latest `openjdk` **22.0.1**?

- #173964

**Pull request merged**
